cmake/multi_image.cmake

```diff
@@ -139,146 +398,20 @@ function(add_child_image_from_source)
-    endif()
-    if(DEFINED ${ACI_NAME}_CONF_FILE)
-      set(${ACI_NAME}_CONF_FILE ${${ACI_NAME}_CONF_FILE} CACHE STRING
-          "Default ${ACI_NAME} configuration file"
-      )
-    endif()
-  endif()
-  # Construct a list of variables that, when present in the root
-  # image, should be passed on to all child images as well.
-  list(APPEND
-    SHARED_MULTI_IMAGE_VARIABLES
-    CMAKE_BUILD_TYPE
-    CMAKE_VERBOSE_MAKEFILE
-    BOARD_DIR
-    BOARD_REVISION
-    ZEPHYR_MODULES
-    ZEPHYR_EXTRA_MODULES
-    ZEPHYR_TOOLCHAIN_VARIANT
-    GNUARMEMB_TOOLCHAIN_PATH
-    EXTRA_KCONFIG_TARGETS
-    NCS_TOOLCHAIN_VERSION
-    PM_DOMAINS
-    ${ACI_DOMAIN}_PM_DOMAIN_DYNAMIC_PARTITION
-    )
-
-  foreach(kconfig_target ${EXTRA_KCONFIG_TARGETS})
-    list(APPEND
-      SHARED_MULTI_IMAGE_VARIABLES
-      EXTRA_KCONFIG_TARGET_COMMAND_FOR_${kconfig_target}
-      )
-  endforeach()
-
   unset(image_cmake_args)
-  list(REMOVE_DUPLICATES SHARED_MULTI_IMAGE_VARIABLES)
-  foreach(shared_var ${SHARED_MULTI_IMAGE_VARIABLES})
-    if(DEFINED ${shared_var})
-      # Any  shared var that is a list must be escaped to ensure correct behaviour.
-      string(REPLACE \; \\\\\; val "${${shared_var}}")
-      list(APPEND image_cmake_args
-        -D${shared_var}=${val}
-        )
-    endif()
-  endforeach()
-
-  get_cmake_property(VARIABLES              VARIABLES)
-  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
-
-  set(regex "^${ACI_NAME}_.+")
-
-  list(FILTER VARIABLES        INCLUDE REGEX ${regex})
-  list(FILTER VARIABLES_CACHED INCLUDE REGEX ${regex})
-
-  foreach(var_name
-      ${VARIABLES}
-      ${VARIABLES_CACHED}
+  if (PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME})
+    # A complete set of configurations has been specified. Use this instead of
+    # finding the cmake args the normal way.
+    set(image_cmake_args "${PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME}}")
+  else()
+    get_child_image_cmake_args(
+      CHILD_IMAGE ${ACI_NAME}
+      CONFIG_OUT image_cmake_args
       )
-    # This regex is guaranteed to match due to the filtering done
-    # above, we only re-run the regex to extract the part after
-    # '_'. We run the regex twice because it is believed that
-    # list(FILTER is faster than doing a string(REGEX on each item.
-    string(REGEX MATCH "^${ACI_NAME}_(.+)" unused_out_var ${var_name})
-
-    # When we try to pass a list on to the child image, like
-    # -DCONF_FILE=a.conf;b.conf, we will get into trouble because ; is
-    # a special character, so we escape it (mucho) to get the expected
-    # behaviour.
-    string(REPLACE \; \\\\\; val "${${var_name}}")
+  endif()
 
-    list(APPEND image_cmake_args
-      -D${CMAKE_MATCH_1}=${val}
-      )
-  endforeach()
+  # Expose the CMake args for the current image in case we want to create a variant build
+  set(${ACI_NAME}_IMAGE_CMAKE_ARGS "${image_cmake_args}" CACHE INTERNAL "")
 
   file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/${ACI_NAME})
   execute_process(
```

Comment:
${EXTRA_MULTI_IMAGE_CMAKE_ARGS} # E.g. --trace-expand

which is below here looks scary.

How do we know people are not putting, e.g.

-DCONFIG_FOO=y

in there and accidentally creating a difference between S0 and S1 ?

Reply:
It is passed down since its a regular CLI arg.